Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -829,10 +829,19 @@ public class ProgramBuilder {
                 self.splice(from: program)
             }, {
                 // We can't run code generators if we don't have any visible variables.
-                guard self.scopeAnalyzer.visibleVariables.count > 0 else { return }
-                let generator = self.fuzzer.codeGenerators.randomElement()
-                if generator.requiredContext.isSubset(of: self.context) {
-                    self.run(generator)
+                if self.scopeAnalyzer.visibleVariables.count == 0 {
+                    // Generate some variables
+                    self.run(self.fuzzer.programPrefixGenerators.randomElement()!)
+                }
+                
+                // Select a generator at random and run it
+                while true {
```

Comment:
I'm a bit paranoid about `while true` since there've been various bugs in the past that have led to infinite loops. Can we do one of these instead:

- Turn this into a for loop that runs, say, 100 times. Turn the `break` in line 843 into a `return`. Log an error after the loop, stating that we couldn't find a suitable CodeGenerator to run in the active context
- Follow the approach suggested in https://github.com/googleprojectzero/fuzzilli/issues/235 and first generate a list of all suitable CodeGenerators, then pick a random one from that list

Reply:
I'm in favour of the first options (i.e. running a loop over fixed iterations). I did consider the proposal in #235 but then we'd be looping over all generators (multiplied by their weights) at each codegen call which is probably slower than just running fewer iterations to pick a random element.

Reply:
Yep, sounds good!

Reply:
On 2nd, I'm not so sure anymore. Let's say we have
```
BeginSwitch
// [ X ]
BeginSwitchCase
...
```
Here, if we try CodeGeneration at point `X`, we would try 50 times, then fail, since we don't have a `SwitchCaseGenerator` (even if we did, the probability of randomly finding it would be pretty low). The more restricted contexts we have, the bigger this issue becomes. So maybe the 2nd approach would actually be better here? We can then just add a comment that we could build a cache of available code generators indexed by context if performance ever becomes an issue.

Reply:
Agreed, a cache would make sense here. For now I'll change this to the second approach.

Reply:
[Don't know why my latest comment disappeared, but I guess you saw it :)]

Reply:
woah thats odd! I saw the comment posted in two places and that was strange but now they both seem to have been removed :D